src/storage/v2/durability/wal.cpp

```diff
@@ -142,50 +143,281 @@ constexpr Marker DeltaActionToMarker(Delta::Action action) {
+    case DELTA_EXISTENCE_CONSTRAINT_DROP:
+    case DELTA_UNIQUE_CONSTRAINT_CREATE:
+    case DELTA_UNIQUE_CONSTRAINT_DROP:
+    case DELTA_ENUM_CREATE:
+    case DELTA_ENUM_ALTER_ADD:
+    case DELTA_ENUM_ALTER_UPDATE:
+    case DELTA_POINT_INDEX_CREATE:
+    case DELTA_POINT_INDEX_DROP:
+    case DELTA_TYPE_CONSTRAINT_CREATE:
+    case DELTA_TYPE_CONSTRAINT_DROP:
+      return true;
+
+    // Not deltas
+    case TYPE_NULL:
+    case TYPE_BOOL:
+    case TYPE_INT:
+    case TYPE_DOUBLE:
+    case TYPE_STRING:
+    case TYPE_LIST:
+    case TYPE_MAP:
+    case TYPE_TEMPORAL_DATA:
+    case TYPE_ZONED_TEMPORAL_DATA:
+    case TYPE_PROPERTY_VALUE:
+    case TYPE_ENUM:
+    case TYPE_POINT_2D:
+    case TYPE_POINT_3D:
+    case SECTION_VERTEX:
+    case SECTION_EDGE:
+    case SECTION_MAPPER:
+    case SECTION_METADATA:
+    case SECTION_INDICES:
+    case SECTION_CONSTRAINTS:
+    case SECTION_DELTA:
+    case SECTION_EPOCH_HISTORY:
+    case SECTION_EDGE_INDICES:
+    case SECTION_OFFSETS:
+    case SECTION_ENUMS:
+    case VALUE_FALSE:
+    case VALUE_TRUE:
+      throw RecoveryFailure("Invalid WAL data!");
+  }
+}
+
+constexpr bool IsMarkerTransactionEnd(const Marker marker, const uint64_t version = kVersion) {
+  if (version < kMetaDataDeltasHaveExplicitTransactionEnd) [[unlikely]] {
+    return IsMarkerImplicitTransactionEndVersion15(marker);
+  }
+  // All deltas are now handled in a transactional scope
+  return marker == Marker::DELTA_TRANSACTION_END;
+}
+
+template <bool is_read>
+auto Decode(utils::tag_type<Gid> /*unused*/, BaseDecoder *decoder, const uint64_t /*version*/)
+    -> std::conditional_t<is_read, Gid, void> {
+  const auto gid = decoder->ReadUint();
+  if (!gid) throw RecoveryFailure("Invalid WAL data!");
+  if constexpr (is_read) {
+    return Gid::FromUint(*gid);
+  }
+}
+
+template <bool is_read>
+auto Decode(utils::tag_type<std::string> /*version*/, BaseDecoder *decoder, const uint64_t /*version*/)
+    -> std::conditional_t<is_read, std::string, void> {
+  if constexpr (is_read) {
+    auto str = decoder->ReadString();
+    if (!str) throw RecoveryFailure("Invalid WAL data!");
+    return *std::move(str);
+  } else {
+    if (!decoder->SkipString()) throw RecoveryFailure("Invalid WAL data!");
+  }
+}
+
+template <bool is_read>
+auto Decode(utils::tag_type<PropertyValue> /*version*/, BaseDecoder *decoder, const uint64_t /*version*/)
+    -> std::conditional_t<is_read, PropertyValue, void> {
+  if constexpr (is_read) {
+    auto str = decoder->ReadPropertyValue();
+    if (!str) throw RecoveryFailure("Invalid WAL data!");
+    return *std::move(str);
+  } else {
+    if (!decoder->SkipPropertyValue()) throw RecoveryFailure("Invalid WAL data!");
+  }
+}
+
+template <bool is_read, auto VER>
+auto Decode(utils::tag_type<VersionDependantGid<VER>> /*version*/, BaseDecoder *decoder, const uint64_t version)
+    -> std::conditional_t<is_read, std::optional<Gid>, void> {
+  if (version <= VER) {
```

Comment:
Good job for tests...it caught this edge case. I've got my inequality the wrong away around.
Only if `version` is `VER` or greater do we do the read/skip.